tlatools/org.lamport.tlatools/test/tla2sany/ParseCorpusTest.java

```diff
@@ -0,0 +1,93 @@
+package tla2sany;
+
+import tla2sany.CorpusParser.CorpusTestFile;
+import tla2sany.CorpusParser.CorpusTest;
+import tla2sany.configuration.Configuration;
+import tla2sany.parser.TLAplusParser;
+import tla2sany.semantic.AbortException;
+import tla2sany.semantic.BuiltInLevel;
+
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.io.InputStream;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.text.ParseException;
+import java.util.List;
+
+/**
+ * Runs all corpus tests through SANY, checking its syntax parsing.
+ */
+public class ParseCorpusTest {
+	
+	/**
+	 * The parsed corpus test files.
+	 */
+	private static List<CorpusTestFile> corpus;
+	
+	/**
+	 * Loads all corpus test files and performs static initialization of SANY.
+	 * 
+	 * @throws IOException If corpus test file cannot be found or read.
+	 * @throws ParseException If corpus test file fails to parse.
+	 * @throws AbortException If SANY static initialization fails.
+	 */
+	@BeforeClass
+	public static void setup() throws IOException, ParseException, AbortException {
+		Path toolsRoot = Paths.get(System.getProperty("basedir"));
+		ParseCorpusTest.corpus = CorpusParser.getAndParseCorpusTestFiles(toolsRoot);
+		Configuration.load(null);
+		BuiltInLevel.load();
+	}
+	
+	/**
+	 * Iterates through each corpus test in each corpus test file, feeds the
+	 * raw TLA+ input into SANY, translates SANY's output to the format
+	 * expected by the test, then compares this translated output to the
+	 * expected parse tree associated with that test.
+	 * 
+	 * @throws ParseException If translating SANY's output fails.
+	 */
+	@Test
+	public void testEntireCorpus() throws ParseException {
+		int testCount = 0;
+		for (CorpusTestFile corpusTestFile : ParseCorpusTest.corpus) {
+			System.out.println(corpusTestFile.path);
+			for (CorpusTest corpusTest : corpusTestFile.tests) {
+				// keeping this here because it is useful to debug why a specific test fails
+				if (!corpusTest.name.equals("Subexpression Tree Navigation")) {
+					//continue;
```

Comment:
?

Reply:
For future development and adding of tests. The main way I translated SANY's parse tree was running the debugger and seeing what SANY emits at a given point in the giant switch statement. Some of the rules are run through hundreds of times if you run the entire corpus, so it's useful to be able to pick out a single test to run so you get the debugger state in a specific rule for that test. That's what this code stub does, if you uncomment continue. It can be removed but it might be useful to the next person (probably me) who adds tests.

Reply:
Added this explanation as a comment in the code.